src/compiler/moduleNameResolver.ts

```diff
@@ -1383,14 +1383,19 @@ namespace ts {
         const onlyRecordFailuresForIndex = onlyRecordFailures || !directoryProbablyExists(candidate, state.host);
         const indexPath = combinePaths(candidate, extensions === Extensions.TSConfig ? "tsconfig" : "index");
 
-        if (versionPaths && (!packageFile || containsPath(candidate, packageFile))) {
-            const moduleName = getRelativePathFromDirectory(candidate, packageFile || indexPath, /*ignoreCase*/ false);
-            if (state.traceEnabled) {
-                trace(state.host, Diagnostics.package_json_has_a_typesVersions_entry_0_that_matches_compiler_version_1_looking_for_a_pattern_to_match_module_name_2, versionPaths.version, version, moduleName);
+        if (versionPaths) {
+            if (!packageFile || containsPath(candidate, packageFile)) {
+                const moduleName = getRelativePathFromDirectory(candidate, packageFile || indexPath, /*ignoreCase*/ false);
+                if (state.traceEnabled) {
+                    trace(state.host, Diagnostics.package_json_has_a_typesVersions_entry_0_that_matches_compiler_version_1_looking_for_a_pattern_to_match_module_name_2, versionPaths.version, version, moduleName);
+                }
+                const result = tryLoadModuleUsingPaths(extensions, moduleName, candidate, versionPaths.paths, loader, onlyRecordFailuresForPackageFile || onlyRecordFailuresForIndex, state);
+                if (result) {
+                    return removeIgnoredPackageId(result.value);
+                }
             }
-            const result = tryLoadModuleUsingPaths(extensions, moduleName, candidate, versionPaths.paths, loader, onlyRecordFailuresForPackageFile || onlyRecordFailuresForIndex, state);
-            if (result) {
-                return removeIgnoredPackageId(result.value);
+            else if (packageFile && !containsPath(candidate, packageFile) && state.traceEnabled) {

```

Comment:
Add state.traceEnabled as first instead of last in && operator to avoid checking the startsWith